src/mjolnir/graphenhancer.cc

```diff
@@ -94,58 +96,232 @@ struct enhancer_stats {
-      speed = static_cast<uint32_t>((speed * kTurnChannelFactor) + 0.5f);
-    } else if ((use == Use::kRamp) && (directededge.speed_type() != SpeedType::kTagged)) {
-      // If no tagged speed set ramp speed to slightly lower than speed
-      // for roads of this classification
-      RoadClass rc = directededge.classification();
-      if ((rc == RoadClass::kMotorway) || (rc == RoadClass::kTrunk) || (rc == RoadClass::kPrimary)) {
-        speed = (density > 8) ? static_cast<uint32_t>((speed * kRampDensityFactor) + 0.5f)
-                              : static_cast<uint32_t>((speed * kRampFactor) + 0.5f);
-      } else {
-        speed = static_cast<uint32_t>((speed * kRampFactor) + 0.5f);
+struct SpeedAssigner {
+  struct SpeedTable {
+    // no special uses
+    std::array<uint32_t, 8> way;
+    // ramps
+    std::array<uint32_t, 6> link_exiting;
+    // turn channel
+    std::array<uint32_t, 6> link_turning;
+    // roundabout
+    std::array<uint32_t, 8> roundabout;
+    // driveway, alley, parking_aisle, drive-through
+    std::array<uint32_t, 4> service;
+
+    template <typename container_t>
+    void parse(const rapidjson::Value& obj, const std::string& name, container_t& entries) {
+      auto arr = obj[name].GetArray();
+      if (arr.Size() != entries.size())
+        throw std::runtime_error(name + " must have " + std::to_string(entries.size()) + " speeds");
+      size_t i = 0;
+      for (const auto& speed : arr) {
+        entries[i++] = speed.GetUint();
       }
     }
-    directededge.set_speed(speed);
 
-    // Done processing links so return...
-    return;
+    SpeedTable(const rapidjson::Value& obj) {
+      parse(obj, "way", way);
+      parse(obj, "link_exiting", link_exiting);
+      parse(obj, "link_turning", link_turning);
+      parse(obj, "roundabout", roundabout);
+      service[0] = obj["driveway"].GetUint();
+      service[1] = obj["alley"].GetUint();
+      service[2] = obj["parking_aisle"].GetUint();
+      service[3] = obj["drive-through"].GetUint();
+    }
+  };
+
+  // 2 letter country and 2 letter state as key with urban and rural speed tables as value
+  std::unordered_map<std::string, std::array<SpeedTable, 2>> tables;
+
+  SpeedAssigner(const boost::optional<std::string>& config_file) {
+    if (!config_file)
+      return;
+    try {
+      auto doc = rapidjson::read_json(*config_file);
+      if (doc.HasParseError())
+        throw std::runtime_error("malformed json");
+      if (!doc.IsArray())
+        throw std::runtime_error("must be a json array");
+
+      // loop over each country/state pair
+      for (const auto& cs : doc.GetArray()) {
+        std::string code = cs["iso3166-1"].GetString();
+        code += cs["iso3166-2"].GetString();
+        tables.emplace(std::move(code), std::array<SpeedTable, 2>{
+                                            SpeedTable(cs["urban"]),
+                                            SpeedTable(cs["rural"]),
```

Comment:
at the moment we have only urban and rural speed tables per each geographic region (country/state combo)